### PR TITLE
Fixed default simulator selection

### PR DESCRIPTION
**Pull request by dsame**
**Problem:** default simulator selection does not work with Xcode 10.1 beta 2.
**Root-cause:** FBSim control returns different data with Xcode 10.1 + old logic can't select iPhone X* simulators.
**Solution:** Align Device Manager with RunLoop simulator retrieving logic.

https://msmobilecenter.visualstudio.com/Mobile-Center/_workitems/edit/50070

### CODE REVIEW

**Comment by jmoody** [comment]
This change set is too complicated.

I want to implement this algorithm:

* https://github.com/calabash/run_loop/blob/develop/lib/run_loop/core.rb#L321

Very easy to understand and easy to update. 

After reviewing the usage of "defaultSimulator", I can image removing the idea of "default simulator" and "default device" entirely.

File: .gitignore [hunk 1]
```diff
@@ -68,3 +68,6 @@ Tests/cucumber/Gemfile*
 
 # Gemfile used in rspec tests for CLI
 Gemfile.lock
+#
+.bundle
+vendor
```
Comment (jmoody) [inline .gitignore]: Why do we need these changes?

I would prefer these were in a different pull request.

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 2]
```diff
@@ -112,39 +112,133 @@ + (NSComparisonResult)comparePreferredSimulator:(FBSimulator *)sim to:(FBSimulat
     NSString *simDeviceName = [sim name];
     NSString *otherSimDeviceName = [otherSim name];
 
-    if ([simVersion isGreaterThan:otherSimVersion]) {
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: Let's implement something less complicated.

**Comment by dsame** [comment]
> I want to implement this algorithm:

I knew about this approach but there's a rub: iOSAgent has no the robust reference to xcode (i suspect it should be "runtime") version.

The subroutine attempts to use run_loop algo:
```
    NSDecimalNumber *simVersion = sim.configuration.os.number;
    NSDecimalNumber *otherSimVersion = otherSim.configuration.os.number;
```
and only if both of simVersion and otherSimVersion are nil fallbacks to regexp approach:

```
 if ((simVersion == nil && otherSimVersion == nil) || [simVersion isEqual:otherSimVersion]) {
 ...
```

I decided to go with fixing RegExp instead of investigating `sim.configuration.os.number` nil case because of RegExp is more predictable.

**Comment by jmoody** [comment]
We can get the Xcode version by shelling out to `xcrun xcodebuild -version` using the ShellRunner + ShellResult.

**Comment by dsame** [comment]
> We can get the Xcode version by shelling out to `xcrun xcodebuild -version` using the ShellRunner + ShellResult.

I got your idea

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 3]
```diff
@@ -1,6 +1,32 @@
 #import "DeviceUtils.h"
 #import "ConsoleWriter.h"
 
+void XcodeVersion(int *iMajor, int* iMinor) {
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: Please use Objective-C methods.

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 4]
```diff
@@ -100,52 +126,47 @@ + (NSString *)findDeviceIDByName:(NSString *)name {
...
+    //   (\\d+|XS)   - pickup either XS or one of 4..8 model and save into capture group #1
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"iPhone\\s+(\\d+|XS)" options:0 error:nil];
+    
+    // while we do not have iOS information will assign each matched simulator
+    // to "def" variable. The last one will be returned
+    FBSimulator *def = nil;
+    for (FBSimulator *simulator in simulators) {
+        NSString *simName = [simulator name];
+        NSArray *matches = [regex matchesInString:simName options:0 range:NSMakeRange(0, [simName length])];
+        if (matches==nil || matches.count == 0) continue;
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: Avoid `nil` checks.

```
// No
if (matches == nil)

// Yes
if (!matches)
```

Don't omit braces. 

```
// No
 if (!matches) continue;

// Yes
 if (!matches)  { continue;}
```

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 5]
```diff
@@ -100,52 +126,47 @@ + (NSString *)findDeviceIDByName:(NSString *)name {
...
+        defaultModel=@"XS";
+    }else{
+        defaultModel=[NSString stringWithFormat:@"%d", xcode_major-1];
+    }
+    
+    // step 2. find FBSimulator with the desired simulators
+    // Re explanation:
+    //   iPhone\\s+  - skip anything before model name
+    //   (\\d+|XS)   - pickup either XS or one of 4..8 model and save into capture group #1
+    NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"iPhone\\s+(\\d+|XS)" options:0 error:nil];
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: Wrap to ~80 characters.

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 6]
```diff
@@ -100,52 +126,47 @@ + (NSString *)findDeviceIDByName:(NSString *)name {
...
+    
+    // while we do not have iOS information will assign each matched simulator
+    // to "def" variable. The last one will be returned
+    FBSimulator *def = nil;
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: Use more descriptive variable names.  I would use `simulator` here.

File: iOSDeviceManager/Utilities/DeviceUtils.m [hunk 7]
```diff
@@ -100,52 +126,47 @@ + (NSString *)findDeviceIDByName:(NSString *)name {
...
+    for (FBSimulator *simulator in simulators) {
+        NSString *simName = [simulator name];
+        NSArray *matches = [regex matchesInString:simName options:0 range:NSMakeRange(0, [simName length])];
+        if (matches==nil || matches.count == 0) continue;
+
+        // rangeAtIndex:0 - the whole match
+        // rangeAtIndex:0 - the first captured group #1
+        NSString *model = [simName substringWithRange:[(NSTextCheckingResult*)matches[0] rangeAtIndex:1]];
+        if ([model isEqualToString:defaultModel]) {
+            def=simulator;
```
Comment (jmoody) [inline iOSDeviceManager/Utilities/DeviceUtils.m]: I would return early here:

```
return simulator;
```

there is no need to continue to iterate.

File: iOSDeviceManager/Utilities/DeviceUtils.m (hunk 6 above)
Comment (dsame) [reply iOSDeviceManager/Utilities/DeviceUtils.m]: renamed as `defaultSimulatorCandidate` in order to emphasize we will not pickup the first matched model but will traverse the list till the most recent iOS

File: iOSDeviceManager/Utilities/DeviceUtils.m (hunk 7 above)
Comment (dsame) [reply iOSDeviceManager/Utilities/DeviceUtils.m]: The idea is to iterate till the last entry to get the recent iOS version.

File: iOSDeviceManager/Utilities/DeviceUtils.m (hunk 2 above)
Comment (dsame) [reply iOSDeviceManager/Utilities/DeviceUtils.m]: replaced with the method used in run_loop

File: iOSDeviceManager/Utilities/XcodeUtils.h [hunk 8]
```diff
@@ -0,0 +1,18 @@
+//
+//  XcodeUtils.h
+//  iOSDeviceManager
+//
+//  Created by Sergey Dolin on 10/10/2018.
```
Comment (AndreyMitsyk) [inline iOSDeviceManager/Utilities/XcodeUtils.h]: I will change it.

**Pull request merged**